Core/Src/monitor.c

```diff
@@ -447,46 +257,141 @@ void vFusingMonitor(void *pv_params)
 		uint8_t fuse_2;
 	} fuse_data;
 
-	bool tsms_status = false;
+	fuse_buf = 0;
+
+	if (read_fuses(pdu, fuses)) {
+		fault_data.diag = "Failed to read fuses";
+		queue_fault(&fault_data);
+	}
+
+	for (fuse_t fuse = 0; fuse < MAX_FUSES; fuse++) {
+		fuse_buf |=
+			fuses[fuse]
+			<< fuse; /* Sets the bit at position `fuse` to the state of the fuse */
+	}
+
+	fuse_data.fuse_1 = fuse_buf & 0xFF;
+	fuse_data.fuse_2 = (fuse_buf >> 8) & 0xFF;
+
+	// reverse the bit order
+	fuse_data.fuse_1 = reverse_bits(fuse_data.fuse_1);
+	fuse_data.fuse_2 = reverse_bits(fuse_data.fuse_2);
+
+	memcpy(fuse_msg.data, &fuse_data, fuse_msg.len);
+	if (queue_can_msg(fuse_msg)) {
+		fault_data.diag = "Failed to send CAN message";
+		queue_fault(&fault_data);
+	}
+}
+
+void steeringio_monitor(steeringio_t *wheel)
+{
+	can_msg_t msg = { .id = 0x680, .len = 8, .data = { 0 } };
+	fault_data_t fault_data = { .id = BUTTONS_MONITOR_FAULT,
+				    .severity = DEFCON5 };
+
+	uint8_t button_1 = !HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_4);
+	uint8_t button_2 = !HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_5);
+	uint8_t button_3 = !HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_6);
+	uint8_t button_4 = !HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_7);
+	uint8_t button_5 = !HAL_GPIO_ReadPin(GPIOC, GPIO_PIN_4);
+	uint8_t button_6 = HAL_GPIO_ReadPin(GPIOC, GPIO_PIN_5);
+	uint8_t button_7 = !HAL_GPIO_ReadPin(GPIOB, GPIO_PIN_0);
+	uint8_t button_8 = !HAL_GPIO_ReadPin(GPIOB, GPIO_PIN_1);
+
+	uint8_t button_data = (button_1 << 7) | (button_2 << 6) |
+			      (button_3 << 5) | (button_4 << 4) |
+			      (button_5 << 3) | (button_6 << 2) |
+			      (button_7 << 1) | (button_8);
+
+	steeringio_update(wheel, button_data);
+
+	/* Set the first byte to be the first 8 buttons with each bit representing the pin status */
+	msg.data[0] = button_data;
+	if (queue_can_msg(msg)) {
+		fault_data.diag = "Failed to send steering buttons can message";
+		queue_fault(&fault_data);
+	}
+}
+
+osThreadId_t data_collection_thread;
+const osThreadAttr_t data_collection_attributes = {
+	.name = "DataCollection",
+	.stack_size = 1000 + (32 * 8 + 128 * 8 + 800 + 64 * 16 + 32 * 16),
+	.priority = (osPriority_t)osPriorityBelowNormal,
+};
+
+void vDataCollection(void *pv_params)
+{
+	data_collection_args_t *args = (data_collection_args_t *)pv_params;
+	mpu_t *mpu = args->mpu;
+	pdu_t *pdu = args->pdu;
+	steeringio_t *wheel = args->wheel;
+	free(args);
+	static const uint8_t delay = 20;
+
+	tsms_reading_mutex = osMutexNew(NULL);
 
 	for (;;) {
-		fuse_buf = 0;
+		read_lv_sense(mpu);
+		osThreadYield();
+		osDelay(delay);
```

Comment:
this is what I tried to explain when I argued against megathreads.  basically rn we are sending lv sense & fuse sense every 100ms and inundating the telemetry when we need it like once every second or two.  But we need to read tsms & steeringio very frequently.  Idk the solution to this but I definitely think it is a problem and will just grow as we add a variety of functions.

Reply:
Maybe we could use periodic timers?

Reply:
I put the data ones in timers so that they only send a CAN message every second. Do you want me to also restrict pedal and TSMS messages?

Reply:
Nah that's OK they are needed for other things lv sense and Fuses are purely telemetry.  If you wanna make everything timer based go ahead, but idk how timer priority works and if it may hurt things.

Reply:
yeah this is what the timers are for...seems like scott handles this fine to me?

Reply:
See my thread in slack regarding this.